src/main/java/core/basesyntax/ArrayList.java

```diff
@@ -1,48 +1,124 @@
+        capacity = DEFAULT_CAPACITY;
+        size = DEFAULT_SIZE;
+    }
+
     @Override
     public void add(T value) {
-
+        growIfArrayFull();
+        array[size++] = value;
     }
 
     @Override
     public void add(T value, int index) {
+        checkBounds(index, true);
+
+        if (index == size) {
+            add(value);
+            return;
+        }
 
+        growIfArrayFull();
+        int elementsToShift = ++size - index - 1;
+        System.arraycopy(array, index, array, index + 1, elementsToShift);
+        array[index] = value;
     }
 
     @Override
     public void addAll(List<T> list) {
+        int listSize = list.size();
+        ensureCapacity(listSize);
+
+        for (int i = 0; i < listSize; i++) {
+            add(list.get(i));
+        }
+    }
 
+    private void ensureCapacity(int listLength) {
+        while (!(capacity > size + listLength)) {
+            grow();
+        }
+    }
+
+    private void growIfArrayFull() {
+        if (size == capacity) {
+            grow();
+        }
+    }
+
+    private void grow() {
+        capacity = (int) (capacity * RESIZE_FACTOR);
+        array = Arrays.copyOf(array, capacity);
     }
 
     @Override
     public T get(int index) {
-        return null;
+        checkBounds(index, false);
+        return (T) array[index];
     }
 
     @Override
     public void set(T value, int index) {
-
+        checkBounds(index, false);
+        array[index] = value;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        checkBounds(index, false);
+        T objectRemoved = (T) array[index];
+        int elementsToShift = size - index - 1;
+        System.arraycopy(array, index + 1, array, index, elementsToShift);
+        array[--size] = null;
+        return objectRemoved;
     }
 
     @Override
     public T remove(T element) {
-        return null;
+        for (int i = 0; i < size; i++) {
+            T targetElement = array[i];
+
+            if ((targetElement == null && element == null)
+                    || (targetElement != null && targetElement.equals(element))) {
+                remove(i);
+                return targetElement;
+            }
+        }
+
+        throw new NoSuchElementException("Element " + element.toString() + " does not exist.");
     }
 
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return size == 0;
+    }
+
+    /* There is conflict during checking bounds when getting or adding
+       a value right after last index, hence the boolean in parameter.
+       When checking with 'true', it won't throw exception if index=size. */
+    private void checkBounds(int index, boolean allowEqualSize) {
+        if (index < 0 || (index > size - 1 && (!allowEqualSize || index != size))) {
+            throw new ArrayListIndexOutOfBoundsException("Index out of bounds.");
+        }
```

Comment:
Using flag-arguments is a bad practice. Better have 2 methods if you need it.
But here you can delete boolean allowEqualSize at all (see my comment to `void add(T value, int index)`)